memstore/ingestion.go

```diff
@@ -363,53 +364,102 @@ func writeBatchRecords(columnDeletions []bool,
 		batch.MaxArrivalTime = upsertBatch.ArrivalTime
 	}
 
-	for _, recordInfo := range records {
-		for col := 0; col < upsertBatch.NumColumns; col++ {
-			columnID, err := upsertBatch.GetColumnID(col)
-			if err != nil {
-				return utils.StackError(err, "Failed to get column id for col %d", col)
-			}
-			if columnDeletions[columnID] {
-				continue
-			}
+	// Instead of traversing row by row, we instead do column by column to avoid making checks on each row.
+	for col := 0; col < upsertBatch.NumColumns; col++ {
+		columnID, err := upsertBatch.GetColumnID(col)
+		if err != nil {
+			return utils.StackError(err, "Failed to get column id for col %d", col)
+		}
+		if columnDeletions[columnID] {
+			continue
+		}
 
-			dataValue, err := upsertBatch.GetDataValue(recordInfo.row, col)
-			if err != nil {
-				return utils.StackError(err, "Failed to get column id for col %d", col)
-			}
+		columnUpdateMode := upsertBatch.columns[col].columnUpdateMode
+		columnMode := upsertBatch.columns[col].columnMode
 
-			columnUpdateMode := upsertBatch.columns[col].columnUpdateMode
-			vectorParty := batch.GetOrCreateVectorParty(columnID, true)
+		// we will skip processing this column if
+		// 1. AllValuesDefault
+		// 2. Its default value is nil
+		// 3. columnUpdate is not UpdateForceOverWriter
+		if columnMode == common.AllValuesDefault && nonNilDefault[columnID] == false && columnUpdateMode != common.UpdateForceOverwrite {
+			continue
+		}
 
-			newValue := &dataValue
-			needToWrite := newValue.Valid
+		if col >= len(upsertBatch.columns) {
+			return utils.StackError(nil, "Column index %d out of range %d", col, upsertBatch.columns)
+		}
 
-			if forUpdate {
-				var oldValue common.DataValue
-				// only read oldValue when mode within add, min, max
-				if columnUpdateMode > common.UpdateOverwriteNotNull {
-					oldValue = vectorParty.GetDataValue(recordInfo.index)
+		vectorParty := batch.GetOrCreateVectorParty(columnID, true)
+		dataType := upsertBatch.columns[col].dataType
+		cmpFunc := common.GetCompareFunc(dataType)
+
+		// check whether the update mode is valid based on data type.
+		forceWrite := false
+		if forUpdate {
+			switch columnUpdateMode {
+			case common.UpdateForceOverwrite:
+				// always update
+				forceWrite = true
+			case common.UpdateWithAddition:
+				fallthrough
+			case common.UpdateWithMin:
+				fallthrough
+			case common.UpdateWithMax:
+				if !common.IsNumeric(dataType) {
+					return utils.StackError(nil, "Unsupported data type %x for column update mode %x", dataType, columnUpdateMode)
 				}
+			}
+		}
```

Comment:
actually AllValuesDefault with forceWrite = false has no need to go beyond this point right?